### DISCUSSION TITLE
Using pgbouncer with sqlalchemy

### DISCUSSION THREAD

**Discussion by radiantone** [sqlalchemy/sqlalchemy#8386] [Usage Questions] [answered]
Hi,
   I'm trying to use pgbouncer and postgres with sqlalchemy. Works pretty well, except one problem I can't seem to figure out. My pgbouncer connections don't seem to close when I close sessions. Eventually, I exhaust the pool size and my clients hang, even though I can see I am calling session.close() there. Is this connection driven by create_engine? 

pgbouncer=# show databases;
   name    |    host    | port | database  | force_user | pool_size | reserve_pool | pool_mode | max_connections | current_connections | paused | disabled 
-----------+------------+------+-----------+------------+-----------+--------------+-----------+-----------------+---------------------+--------+----------
 pgbouncer |            | 6432 | pgbouncer | pgbouncer  |         2 |            0 | statement |             100 |                   0 |      0 |        0
 pyfi      | postgresdb | 5432 | pyfi      | postgres   |       100 |            0 |           |             100 |                  35 |      0 |        0
(2 rows)

My clients currently cycle on a timed loop and have maybe a few threads each that open and close sessions to perform searches and modify objects.  Here is my contexthandler they use. Any tips appreciated. 

_engine = create_engine(CONFIG.get("database", "uri"), isolation_level='AUTOCOMMIT')

@contextmanager
def get_session(**kwargs):

    logging.debug("get_session: Creating session")

    session = scoped_session(sessionmaker(bind=_engine))

    try:
        logging.debug("get_session: Yielding session")
        yield session
    except:
        logging.debug("get_session: Rollback session")
        session.rollback()
    else:
        logging.debug("get_session: Commit session")
        session.commit()
    finally:
        logging.debug("get_session: Closing session")
        session.expunge_all()
        session.close()

**Comment by zzzeek** [comment, marked as answer]
use [NullPool](https://docs.sqlalchemy.org/en/14/core/pooling.html?highlight=nullpool#switching-pool-implementations) to start with so that your application is not holding onto any connections.  after that, whatever isn't working is on the pgbouncer side.

**Comment by radiantone** [comment]
Yeah, went that route and then posted on pgbouncer discussions about it. Seems even when my client terminates, it shows active connections

**Reply by zzzeek** [reply]
you mean "idle in transaction" or similar , right ?  it's normal that pgbouncer would have connections opened to the server, their state would be "idle".  after all, it's a connection pool.

**Reply by radiantone** [reply]
Yeah, I guess as much, but it seems its the connections INTO pgbouncer that don't seem to recycle and eventually get exhausted. Even though my clients are closing connections. This strikes me as very counterintuitive. For example, if my pgbouncer has max 20 connections and each of my sqlalchemy clients open 15 connections (closing them when done at different times), I would expect pgbouncer to manage 2 clients easily, sharing the pooled connections on its side. But I don't see this behavior. On top of that the connection count goes up, if a client terminates and then restarts, opening new connections. So far as I can tell, pgbouncer doesn't do anything useful.

So going back to regular postgres everything works fine and when my sqlalchemy connections close, the database sees them as closed and the server side resources are freed.

**Reply by zzzeek** [reply]
> But I don't see this behavior.

what behavior do you see?   is it rejecting connections?

**Reply by radiantone** [reply]
Once the max connections are exceeded the clients all block on pgbouncer and eventually time out. And they close all their connections too. Its as if pgbouncer isnt keeping track of whether a client has closed a connection so it can reuse it. Which is the basic premise behind it even.

**Reply by zzzeek** [reply]
this is after your client program actually exits ?

**Reply by radiantone** [reply]
After client exits pgbouncer still shows "active connections" but under show clients pgbouncer shows none. It really needs to discern between pooled connections that are NOT connected to a client (i.e. available for use) and ones that are. But it doesnt make this distinction so its rather useless otherwise.

**Reply by radiantone** [reply]
When a client exits or closes a connection i expect to see the ACTIVE connection count to drop. To me active means a client is connected to it.

**Reply by zzzeek** [reply]
OK so you need to work with the raw driver and pgbouncer for a bit and then email their help lists to understand what's going on.

**Reply by radiantone** [reply]
I did do that. :)

**Comment by Lewiscowles1986** [comment]
Can I ask on this, if setting a poolsize of 1 is the same as NullPool for behavior? I'm seeing slower session.close() times after a single query than the connection time, and I believe this to mean that SQLAlchemy is somehow involved in slow session.close's But I can't for the life of me work-out how to observe that without adding more risky behavior, all of which will have overhead to sqlalchemy.

If it helps, the behavior is observed within an AWS lamdba using RDS Read only proxy aurora, with 40 of their ACU's across 5 read-replicas (analogous to multi-reader MySQL). I Believe their X-Ray traces hook into pymysql, rather than SQLAlchemy itself; but I just can't reconcile 40ms block for contextmanager getting a session, with only 4-9ms query time otherwise.

I Have experimented with using connection rather than session, and besides breaking a bunch of tests, the workaround code, wound up taking even longer 100's of ms. I'd love to be able to eliminate SQLAlchemy as a possible cause.

**Reply by zzzeek** [reply]
> Can I ask on this, if setting a poolsize of 1 is the same as NullPool for behavior?

no, not at all, those do completely different things

> I'm seeing slower session.close() times after a single query than the connection time, and I believe this to mean that SQLAlchemy is somehow involved in slow session.close's But I can't for the life of me work-out how to observe that without adding more risky behavior, all of which will have overhead to sqlalchemy.

you would want to create a test that you can run in a test environment that either /or uses wallclock time with time.time() / timeit etc. or python profiling to see what is happening.   see https://docs.sqlalchemy.org/en/20/faq/performance.html#how-can-i-profile-a-sqlalchemy-powered-application

> If it helps, the behavior is observed within an AWS lamdba using RDS Read only proxy aurora, with 40 of their ACU's across 5 read-replicas (analogous to multi-reader MySQL). I Believe their X-Ray traces hook into pymysql, rather than SQLAlchemy itself; but I just can't reconcile 40ms block for contextmanager getting a session, with only 4-9ms query time otherwise.

well that's a datapoint since this thread is about postgresql, mysql specifically can have slow "rollback" times which is something the connection pool has to do when a connection is returned to the pool , if you didn't do a commit.   if this DB is read only, then use autocommit mode for your connection (I would set it for the engine overall), that will be significantly faster.  see https://docs.sqlalchemy.org/en/20/core/connections.html#setting-transaction-isolation-levels-including-dbapi-autocommit

**Reply by Lewiscowles1986** [reply]
Thanks, I did sadly observe NodeJS eating my lunch when grabbing the queries out of SQLAlchemy and running parametrized.

Half tempted to make python call NodeJS for all the queries, but will stave off that madness to continue the suggestions. 🙇 

I'll leave this thread now. I did notice it was Postgres, but I was hoping for the exact type of thing you've said. I'll look to do less.
